SignalServiceKit/src/Profiles/OWSProfilesManager.m

```diff
@@ -375,32 +436,164 @@ - (void)updateProfileOnService:(nullable NSString *)localProfileName
+{
+    OWSAssert(recipientId.length > 0);
+
+    [self.dbConnection setObject:@(1) forKey:recipientId inCollection:kOWSProfilesManager_UserWhitelistCollection];
+    self.userProfileWhitelistCache[recipientId] = @(YES);
+}
+
+- (BOOL)isUserInProfileWhitelist:(NSString *)recipientId
 {
-    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
-        NSString *_Nullable localProfileName = [self.dbConnection objectForKey:kOWSProfilesManager_LocalProfileNameKey
-                                                                  inCollection:kOWSProfilesManager_Collection];
-        AvatarMetadata *_Nullable localProfileAvatarMetadata =
-            [self.dbConnection objectForKey:kOWSProfilesManager_LocalProfileAvatarMetadataKey
-                               inCollection:kOWSProfilesManager_Collection];
-        UIImage *_Nullable localProfileAvatarImage = nil;
-        if (localProfileAvatarMetadata) {
-            localProfileAvatarImage = [self loadProfileAvatarWithFilename:localProfileAvatarMetadata.fileName];
-            if (!localProfileAvatarImage) {
-                localProfileAvatarMetadata = nil;
-            }
+    OWSAssert(recipientId.length > 0);
+
+    NSNumber *_Nullable value = self.userProfileWhitelistCache[recipientId];
+    if (value) {
+        return [value boolValue];
+    }
+
+    value =
+        @(nil != [self.dbConnection objectForKey:recipientId inCollection:kOWSProfilesManager_UserWhitelistCollection]);
+    self.userProfileWhitelistCache[recipientId] = value;
+    return [value boolValue];
+}
+
+- (void)addGroupIdToProfileWhitelist:(NSData *)groupId
+{
+    OWSAssert(groupId.length > 0);
+
+    NSString *groupIdKey = [groupId hexadecimalString];
+    [self.dbConnection setObject:@(1) forKey:groupIdKey inCollection:kOWSProfilesManager_GroupWhitelistCollection];
+    self.groupProfileWhitelistCache[groupIdKey] = @(YES);
+}
+
+- (BOOL)isGroupIdInProfileWhitelist:(NSData *)groupId
+{
+    OWSAssert(groupId.length > 0);
+
+    NSString *groupIdKey = [groupId hexadecimalString];
+    NSNumber *_Nullable value = self.groupProfileWhitelistCache[groupIdKey];
+    if (value) {
+        return [value boolValue];
+    }
+
+    value =
+        @(nil != [self.dbConnection objectForKey:groupIdKey inCollection:kOWSProfilesManager_GroupWhitelistCollection]);
+    self.groupProfileWhitelistCache[groupIdKey] = value;
+    return [value boolValue];
+}
+
+- (void)setContactRecipientIds:(NSArray<NSString *> *)contactRecipientIds
+{
+    OWSAssert(contactRecipientIds);
+    
+    // TODO: The persisted whitelist could either be:
+    //
+    // * Just users manually added to the whitelist.
+    // * Also include users auto-added by, for example, being in the user's
+    //   contacts or when the user initiates a 1:1 conversation with them, etc.
+    for (NSString *recipientId in contactRecipientIds) {
+        [self addUserToProfileWhitelist:recipientId];
```

Comment:
As written, this will create a one-write transaction for every one of my contacts (on the main thread), right? 

A couple of perf thoughts:

1. We should be able to do this in a single transaction
2. We should only write to the db when the value differs from what's in our cache (maybe?)